### PR TITLE
Replace taskGraph.afterTask with an operation completion listener in Gradle 6.1+

### PR DESCRIPTION
**Pull request by Vampire**

### CODE REVIEW

**Comment by yorlov** [comment]
Just to let you know, the plugin will not function on `gradle 8.7` until this issue is rectified.

```
FAILURE: Build failed with an exception.

* What went wrong:
Configuration cache problems found in this build.

1 problem was found storing the configuration cache.
- Plugin 'net.researchgate.release': registration of listener on 'TaskExecutionGraph.afterTask' is unsupported
  See https://docs.gradle.org/8.6/userguide/configuration_cache.html#config_cache:requirements:build_listeners
```

**Comment by Vampire** [comment]
You mean the plugin will not function with 8.x if you enable configuration cache, which is not different in 8.7 from earlier versions.

File: src/main/groovy/net/researchgate/release/ReleasePlugin.groovy [hunk 1]
```diff
@@ -157,18 +164,40 @@ class ReleasePlugin extends PluginHelper implements Plugin<Project> {
...
+    abstract protected ObjectFactory getObjects();
+
+    @Inject
+    abstract protected ProviderFactory getProviders();
+
+    protected revert() {
+        try {
+            createScmAdapter()
+        } catch (Exception ignored) {}
+        if (scmAdapter && extension.revertOnFail && project.file(extension.versionPropertyFile)?.exists()) {
```
Comment (Hillkorn) [inline src/main/groovy/net/researchgate/release/ReleasePlugin.groovy:196]:
```suggestion
        if (scmAdapter && extension.revertOnFail.get() && project.file(extension.versionPropertyFile)?.exists()) {
```
Not sure what groovy does if we don't use the get() .

File: src/main/groovy/net/researchgate/release/ReleasePlugin.groovy [hunk 2]
```diff
@@ -28,12 +28,19 @@ import org.gradle.api.GradleException
...
 import org.gradle.api.tasks.TaskState
+import org.gradle.build.event.BuildEventsListenerRegistry
+import org.gradle.tooling.events.OperationCompletionListener
+import org.gradle.tooling.events.task.TaskFailureResult
+import org.gradle.util.GradleVersion
 
-class ReleasePlugin extends PluginHelper implements Plugin<Project> {
+import javax.inject.Inject
 
+abstract class ReleasePlugin extends PluginHelper implements Plugin<Project> {
```
Comment (Hillkorn) [inline src/main/groovy/net/researchgate/release/ReleasePlugin.groovy:43]: What is the reason for this change?
I see that you added
```
    @Inject
    abstract protected ObjectFactory getObjects();

    @Inject
    abstract protected ProviderFactory getProviders();
 ```
But I don't get the why.

Comment (Vampire) [reply src/main/groovy/net/researchgate/release/ReleasePlugin.groovy:43]: Because I use them at https://github.com/researchgate/gradle-release/blob/7ffe081f69545e66ad4625ee5ed68e7714c44b36/src/main/groovy/net/researchgate/release/ReleasePlugin.groovy#L174 and https://github.com/researchgate/gradle-release/blob/7ffe081f69545e66ad4625ee5ed68e7714c44b36/src/main/groovy/net/researchgate/release/ReleasePlugin.groovy#L176

Of course you can get those from `project` too, but that is bad / outdated practice and I didn't want to add to the existing ones but doing it right. :-)

File: src/main/groovy/net/researchgate/release/ReleasePlugin.groovy (hunk 1 above)
Comment (Vampire) [reply src/main/groovy/net/researchgate/release/ReleasePlugin.groovy]: The same it does currently, I just moved this existing code to a method.
But you are fully right of course, for Groovy it will just always be `true` as it is non-null.
This is already like that since switching to properties and not adapting this usage (and maybe others?)
One of the fun things you have because of using Groovy for the plugin. :-D

I now fixed it together with this PR.

**Review by Hillkorn — APPROVED** [review]

**Comment by daniel-shuy** [comment]
@Vampire Thanks for this PR!

@Hillkorn any idea when can we see this released?
